genericjpa/async-updates/src/main/java/org/hibernate/search/db/events/jpa/impl/SQLJPAAsyncUpdateSourceProvider.java

```diff
@@ -0,0 +1,204 @@
+				this.triggerCreateStrategy
+		)) ) {
+			return;
+		}
+
+		try {
+			try {
+				if ( AsyncUpdateConstants.TRIGGER_CREATION_STRATEGY_DROP_CREATE.equals( this.triggerCreateStrategy ) ) {
+					//DROP EVERYTHING IN THE EXACTLY INVERSED ORDER WE CREATE IT
+					for ( EventModelInfo info : eventModelInfos ) {
+
+						for ( int eventType : EventType.values() ) {
+							String[] triggerDropStrings = this.triggerSource.getTriggerDropCode( info, eventType );
+							for ( String triggerDropString : triggerDropStrings ) {
+								log.trace( triggerDropString );
+								this.doQueryOrLogException(
+										emf,
+										triggerDropString,
+										true
+								);
+							}
+						}
+
+						for ( String unSetupCode : this.triggerSource.getSpecificUnSetupCode( info ) ) {
+							log.trace( unSetupCode );
+							this.doQueryOrLogException( emf, unSetupCode, true );
+						}
+
+						for ( String str : triggerSource.getUpdateTableDropCode( info ) ) {
+							log.trace( str );
+							this.doQueryOrLogException( emf, str, true );
+						}
+
+					}
+
+					for ( String str : triggerSource.getUnSetupCode() ) {
+						log.trace( str );
+						this.doQueryOrLogException( emf, str, true );
+					}
+				}
+
+				//CREATE EVERYTHING
+				try {
+					for ( String str : triggerSource.getSetupCode() ) {
+						log.trace( str );
+						this.doQueryOrLogException( emf, str, false );
+					}
+
+					for ( EventModelInfo info : eventModelInfos ) {
+						for ( String str : triggerSource.getUpdateTableCreationCode( info ) ) {
+							log.trace( str );
+							this.doQueryOrLogException( emf, str, false );
+						}
+
+						for ( String setupCode : this.triggerSource.getSpecificSetupCode( info ) ) {
+							log.trace( setupCode );
+							this.doQueryOrLogException( emf, setupCode, false );
+						}
+
+						for ( int eventType : EventType.values() ) {
+							String[] triggerCreationStrings = this.triggerSource.getTriggerCreationCode(
+									info,
+									eventType
+							);
+							for ( String triggerCreationString : triggerCreationStrings ) {
+								log.trace( triggerCreationString );
+								this.doQueryOrLogException(
+										emf,
+										triggerCreationString,
+										false
+								);
+							}
+						}
+					}
+				}
+				catch (Exception e) {
+					throw new SearchException( e );
+				}
+				log.trace( "finished setting up triggers!" );
+			}
+			finally {
+
+			}
+		}
+		catch (Exception e) {
+			throw new SearchException( e );
+		}
+	}
+
+	private void doQueryOrLogException(
+			EntityManagerFactoryWrapper emf,
+			String query,
+			boolean canFail) {
+		try {
+			//we use a new EntityManager here everytime, because
+			//if we get an error during trigger creation
+			//(which is allowed, since we don't have logic
+			//to check with IF EXISTS on every database)
+			//the EntityManager can be in a RollbackOnly state
+			//which we dont want
+			EntityManagerWrapper em = emf.createEntityManager();
```

Comment:
I cannot unwrap to a JDBC connection from every JPA provider. (I believe Hibernate is the one that restricts this)